man/man1/pmnewlog.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMNEWLOG 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmnewlog\f1 \- stop and restart archive logging for PCP performance metrics
.SH SYNOPSIS
\f3$PCP_BINADM_DIR/pmnewlog\f1
[\f3\-NPsV?\f1]
[\f3\-a\f1 \f2accessfile\f1]
[\f3\-c\f1 \f2configfile\f1]
[\f3\-C\f1 \f2saveconfig\f1]
[\f3\-n\f1 \f2pmnsfile\f1]
[\f3\-p\f1 \f2pid\f1]
[\f2other pmlogger options\f1]
\f2archive\f1
.SH DESCRIPTION
.B pmnewlog
may be used to stop and restart a running instance of
.BR pmlogger (1).
This is most useful for managing multiple sets of
Performance Co-Pilot (PCP) archive logs.
These archive logs record the history of
performance metric values
that may be ``played back'' by other PCP
tools, and they
form the basis of the VCR paradigm and retrospective
performance analysis services common to the PCP toolkit.
.PP
In normal usage,
.B pmnewlog
would be executed by
.BR cron (1)
in the wee hours to terminate one PCP archive log and start another,
i.e. to perform log rotation.
.PP
Even more common, would be the execution of
.B pmnewlog
from the PCP archive management script
.BR pmlogger_daily (1).
In this case, direct end-user execution of
.B pmnewlog
is most unlikely.
.PP
The mandatory argument
.I archive
is the base name for the physical files that will constitute
the new archive log.
.PP
The
.B pmlogger
instance to be stopped and restarted must be running on the same system
as
.B pmnewlog
and is either the primary logger (the default) or the logger with
.I pid
as specified by the
.B \-p
option.
.PP
If the
.B \-n
option is specified, then
.B pmnewlog
will use the namespace in the
.IR pmnsfile ,
rather than the default Performance Metrics Name Space (PMNS).
.PP
If no
.B \-c
option is specified,
.B pmnewlog
will use
.BR pmlc (1)
to connect to the running
.BR pmlogger (1)
and so determine all those metrics and instances that are subject to
.B mandatory
logging or
.B advisory on
logging, and the associated logging frequencies.
This information is used to synthesize a new
.BR pmlogger (1)
configuration file.
If the
.B \-n
option is specified, it will also be used for these interactions with
.BR pmlc (1).
.PP
If the
.B \-c
option is specified,
.BR pmlogger (1)
will be restarted with
.I configfile
as the configuration file.
Normally
.I configfile
would be the same configuration file used to start
.BR pmlogger (1)
in the first place, however note that since
.BR pmlogger (1)
is restarted, any changes to the logging status made
using
.BR pmlc (1)
will be lost, unless these have also been reflected in changes to
.IR configfile .
.PP
If
.I configfile
does not exist, then a search is made in the directory
.I $PCP_VAR_DIR/config/pmlogger
for a file of the same name, and if found that file is used,
e.g. if
.I config.mumble
does not exist in the current directory and
the file
.I $PCP_VAR_DIR/config/pmlogger/config.mumble
does exist, then
.B "\-c config.mumble"
and
.B "\-c $PCP_VAR_DIR/config/pmlogger/config.mumble"
are equivalent.
.PP
Access controls specifications for the new
.BR pmlogger (1)
instance may optionally be provided via the
.B \-a
option.
The contents of
.I accessfile
should start with the literal token
.B [access]
and conform to the syntax of the access controls section
as described for
.BR pmlogger (1).
.PP
The
.B \-C
option may be used to save the configuration file that
.B pmnewlog
passes to the newly launched
.BR pmlogger (1).
.PP
If the
.BR pmlogger (1)
instance needs to be started under the control of
.BR pmsocks (1)
to connect to a
.B pmcd
through a firewall, the
.B \-s
option may be used.
.PP
The
.B \-V
option enables verbose reporting of the activity.
By default no output is generated unless some error or warning condition is
encountered.
.PP
The
.B \-N
option enables a ``show me'' mode, where the actions are echoed,
but not executed, in the style of ``make \-n''.
Using
.B \-N
in conjunction with
.B \-V
maximizes the diagnostic capabilities for debugging.
.PP
The
.I other pmlogger options
are as described for
.BR pmlogger (1).
Note that
.B pmnewlog
does
.B not
support the following options of
.BR pmlogger (1).
.TP
\fB\-h\fR \fIhost\fR
.B pmnewlog
determines the host to which the new
.BR pmlogger (1)
should connect based upon the current host connection for the
old
.BR pmlogger (1).
.TP
\fB\-s\fR \fIsamples\fR
The new
.BR pmlogger (1)
is expected to be long running, and the
.B \-s
option of
.B pmnewlog
takes precedence.
.TP
\fB\-T\fR \fIruntime\fR
The new
.BR pmlogger (1)
is expected to be long running.
.TP
\fB\-V\fR \fIversion\fR
The new
.B pmlogger
will always create the latest version PCP archive format, and the
.B \-V
option of
.B pmnewlog
takes precedence.
.TP
\fB\-x\fR \fIfd\fR
The launched
.B pmlogger
cannot be controlled by
.BR pmRecordControl (3).
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-a\fR \fIfile\fR, \fB\-\-access\fR=\fIfile\fR
Specify access controls \fIfile\fR for the new pmlogger.
.TP
\fB\-c\fR \fIfile\fR, \fB\-\-config\fR=\fIfile\fR
Load configuration from \fIfile\fR.
.TP
\fB\-C\fR \fIfile\fR, \fB\-\-save\fR=\fIfile\fR
Save the configuration of new pmlogger in \fIfile\fR.
.TP
\fB\-n\fR \fIpmnsfile\fR, \fB\-\-namespace\fR=\fIpmnsfile\fR
Load an alternative Performance Metrics Name Space
.RB ( PMNS (5))
from the file
.IR pmnsfile .
.TP
\fB\-N\fR, \fB\-\-showme\fR
Perform a dry run.
.TP
\fB\-p\fR \fIPID\fR, \fB\-\-pid\fR=\fIPID\fR
Restart non-primary logger with PID \fIPID\fR.
.TP
\fB\-P\fR, \fB\-\-primary\fR
Execute as primary logger instance.
.TP
\fB\-s\fR, \fB\-\-socks\fR
Use
.BR pmsocks (1)
to connect.
.TP
\fB\-V\fR, \fB\-\-verbose\fR
Use verbose reporting.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH EXAMPLES
The following
.BR sh (1)
script
could be executed by root via
.BR cron (1)
to start a new set of archive logs for the primary logger each evening.
A more complete version of this script may be found in
.IR $PCP_BINADM_DIR/pmlogger_daily ,
and is documented in the manual page for
.BR pmlogger_daily (1).
.PP
.in +8n
.nf
.ft CW
#!/bin/sh
# start new logs for PCP primary logger on this host

# standard place for logs
LOGDIR=$PCP_LOG_DIR/pmlogger/`hostname`

# each new log is named yymmdd.hh.mm
LOGNAME=`date "+%Y%m%d.%H.%M"`

# do it
[ ! \-d $LOGDIR ] && mkdir \-p $LOGDIR
cd $LOGDIR
$PCP_BINADM_DIR/pmnewlog \-l $LOGDIR/pmlogger.log $LOGDIR
.ft R
.fi
.in -8n
.SH CAVEATS
If no
.I configfile
is specified, the method for synthesizing a configuration file using
a
.BR pmlc (1)
connection to the existing
.BR pmlogger (1)
is, of necessity, incomplete.
In particular,
for metrics with dynamic underlying instance domains,
it is not possible to identify a configuration that logs
.B all
instances of a metric all of the time,
so rather the synthesized configuration file requests the continued logging
of the set of instances that exist at the time
.BR pmlogger (1)
is interrogated by
.BR pmnewlog .
.PP
If this situation is a concern, a fixed configuration file should
be used, and passed to
.B pmnewlog
via the
.B \-c
option.
.SH DIAGNOSTICS
Due to the precious nature of the archive logs,
.B pmnewlog
is rather paranoid in its checking and validation, and will try very
hard to ensure that an appropriately configured
.BR pmlogger (1)
can be restarted, before terminating the existing
.BR pmlogger (1).
.PP
As a consequence of this checking,
.B pmnewlog
tends to generate rather verbose error and warning messages.
.SH FILES
.TP 5
\f2archive\f3.meta\f1
metadata (metric descriptions, instance domains, etc.) for the archive log
.TP
\f2archive\f3.0\f1
initial volume of metrics values (subsequent volumes have suffixes
.BR 1 ,
.BR 2 ,
\&...)
.TP
\f2archive\f3.index\f1
temporal index to support rapid random access to the other files in the
archive log
.TP
.I $PCP_BINADM_DIR/pmlogger_daily
sample script to rotate archives for a number of loggers
.TP
.B SaveLogs
if this directory exists within the directory that the
.I archive
files will be created by a new
.BR pmlogger (1)
then the log file (from
.BR pmlogger 's
.B \-l
argument) will be linked into the
.B SaveLogs
directory with the name
.IB archive .log
so it can be inspected at a later time.
Because the cron-driven PCP archive management scripts run under
the uid of the user ``pcp'',
.B SaveLogs
typically needs to be owned by the user ``pcp''.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize

CHANGELOG
pcp-5.0.3 (27 February 2020)
    - Work in progress: https://github.com/performancecopilot/pcp/projects/1

pcp-5.0.2 (11 December 2019)
    - Server-side utilities and log management scripts:
	pmproxy: REST API to report all label values by name
	pmproxy: fix memory leak when re-using http connections
	pmproxy: fix memory leak on a failed REST API call
	pmproxy: improvements to the instrumentation mechanism
	pmproxy: add support for AF_UNIX local client connections
	logutil: fix pmlogger service timeout on new installations
	pmie: on SIGTERM, exit 0 for a normal shutdown rather than exit 15
	pmcd: add labels PDUs into the trace set
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: support for kernel cgroups v2 (and systemd using same)
	pmdalinux: update per-device read_bytes, write_bytes metric types
	pmdabpftrace: implement label callback
	pmdabpftrace: clear context tab on endcontext
	pmdanetcheck: run as pcp user, document ICMP Echo socket permissions
	pmdaperfevent: fix memory leaks and remove some unused code
    - Client tools and utilities:
	pcp-atop: update to latest upstream atop v2.5.0 sources
	pcp-dstat: no stack trace when dstat exits due to a signal
	pcp-dstat: fix handling of very large numbers of disk devices
	pcp-dstat: exit with code zero when archive end reached
	pmieconf: add a pcp-zeroconf option for enabling all_threads
	pmseries: fix pointer arithmetic on series ID set intersection
	pmseries: add new option to report all label values by name
	pmseries: fix and test regular expression pattern matching
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings
	libpcp: fix potential use-after-free in labels code
	libpcp_web: fix file descriptor and memory leaks in discovery code
	libpcp_web: resolve valgrind issues observed in qa tests
	libpcp_pmda: show PM_ERR_VALUE errors only if libpmda debug flag set
	python API: add endcontext callback for PMDAs
	python API: resolve pylint warnings on all python modules
	python API: add missing fields to pmLabelSets structure
    - Misc build, infrastructure and packaging updates:
	configure: make build reproducible
	configure: make zlib available for hdr_histogram checks
	build: do not create dstat symlink when non-python build requested
	build: add perl-High-Res rpm dependency for pcp-pmda-postfix
	build: update deb packaging to allow for python3-only builds
	build: resolve several rpm spec differences that have crept in
	build: fix chan lib dependency of pcp-pmda-statsd rpm
	build: fix hostname dependency of pcp rpm
    - Security Enhanced Linux:
	pmdaunbound: add selinux policy for unbound PMDA metrics
	pmdanetcheck: resolve failure on latest f31
	pmdabpftrace: resolve failure on latest f31
	pmdabcc: fix AVC error when running /var/lib/pcp/pmdas/bcc/Install
	pmie: add fsetid to pmie start script policy as done for pmlogger
    - Documentation and QA infrastructure:
	CI: initial new version using custom Azure Pipelines agents and Ansible
	CI: simplified architecture, added pipeline definitions
	CI: create JUnit test results file for Azure Pipelines
	CI: run platform builds and tests in parallel
	docs: Added to pmseries(1) man page
	docs: Added Redis setup notes to the quickstart guide
	docs: Complete coverage in man pages of long options in PCP utilities.
	docs: Numerous man page consistency fixes
	docs: details for pmSeriesSetup(3) and related async APIs
	docs: details for pmDiscoverSetup(3) and related async APIs
	docs: details for pmSeriesQuery(3) and related time series APIs
	docs: details for pmSeriesDescs(3) and other time series metadata APIs

pcp-5.0.1 (4 November 2019)
    - Server-side utilities and log management scripts:
	pmproxy: fix bad memory accesses reported by valgrind
	pmproxy: guard libuv socket close to prevent double-close failure
	pmproxy: ensure client data structures are freed only on last use
	pmproxy: remove linefeeds at the end of JSON responses
	pmproxy: ensure correct persistent connection streaming state
	pmproxy: ensure uv_write happens from main thread
	pmproxy: refresh indom if instance doesn't exist in the cached indom
	pmproxy: always refresh indom for /indom API requests
	pmproxy: improve robustness parsing URL with unusual params
    - PMDA additions, enhancements and bug fixes:
	pmdalinux: fix segfault in interrupts code with large CPU counts
	pmdalmsensors: fix issue when sensors(1) writes on stderr
	pmdanetcheck: netcheck.conf and man page consistency tweaks

src/pmlogger/pmnewlog.sh
#! /bin/sh
#
# Copyright (c) 2014,2018 Red Hat.
# Copyright (c) 1995-2001,2003 Silicon Graphics, Inc.  All Rights Reserved.
# 
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# stop and restart a pmlogger instance
#

# Get standard environment
. $PCP_DIR/etc/pcp.env

PMLOGGER="$PCP_BINADM_DIR/pmlogger"
PMLOGGERENVS="$PCP_SYSCONFIG_DIR/pmlogger"

# error messages should go to stderr, not the GUI notifiers
#
unset PCP_STDERR

tmp=`mktemp -d /tmp/pcp.XXXXXXXXX` || exit 1
status=0
trap "rm -rf $tmp; exit \$status" 0 1 2 3 15
prog=`basename $0`

VERBOSE=false
SHOWME=false
CP=cp
MV=mv
RM=rm
KILL=pmsignal
primary=true
myname="primary pmlogger"
connect=primary
access=""
config=""
saveconfig=""
logfile="pmlogger.log"
namespace=""
args=""
sock_me=""
proxyhost="$PMPROXY_HOST"
proxyport="$PMPROXY_PORT"

cat > $tmp/usage << EOF
# Usage: [options] archive

pmnewlog options:
  -a=FILE, --access=FILE  specify access controls for the new pmlogger
  -C=FILE, --save=FILE    save the configuration of new pmlogger in FILE
  -c=FILE, --config=FILE  file to load configuration from
  -N, --showme            perform a dry run, showing what would be done
  -n=FILE, --namespace=FILE use an alternative PMNS
  -P, --primary           execute as primary logger instance
  -p=PID, --pid=PID       restart non-primary logger with pid
  -s, --socks             use pmsocks
  -V, --verbose           turn on verbose reporting of pmnewlog progress
  --help

pmlogger options:
  --debug
  -c=FILE, --config=FILE  file to load configuration from
  -H=LABELHOST, --labelhost override the hostname written into the label
  -l=FILE, --log=FILE     redirect diagnostics and trace output
  -L, --linger            run even if not primary logger instance and nothing to log
  -m=MSG, --note=MSG      descriptive note to be added to the port map file
  -n=FILE, --namespace=FILE use an alternative PMNS
  -K=SPEC, --spec-local=SPEC optional additional PMDA spec for local connection
  -o, --local-PMDA        metrics source is local connection to a PMDA
  -P, --primary           execute as primary logger instance
  -r, --report            report record sizes and archive growth rate
  -t=DELTA, --interval=DELTA default logging interval
  -T=TIME, --finish=TIME  end of the time window
  -v=SIZE, --volsize=SIZE switch log volumes after size has been accumulated
  -y                      set timezone for times to local time rather than from PMCD host
EOF

_abandon()
{
    echo
    echo "Sorry, but this is fatal.  No new pmlogger instance has been started."
    status=1
    exit
}

_check_pid()
{
    if $SHOWME
    then
	:
    else
	_get_pids_by_name pmlogger | grep "^$1\$"
    fi
}

_check_logfile()
{
    if [ ! -f $logfile ]
    then
	echo "Cannot find pmlogger output file at \"$logfile\""
    else
	echo "Contents of pmlogger output file \"$logfile\" ..."
	cat $logfile
    fi
}

_check_logger()
{
    # wait until pmlogger process starts, or exits
    #
    delay=5
    [ ! -z "$PMCD_CONNECT_TIMEOUT" ] && delay=$PMCD_CONNECT_TIMEOUT
    x=5
    [ ! -z "$PMCD_REQUEST_TIMEOUT" ] && x=$PMCD_REQUEST_TIMEOUT

    # wait for maximum time of a connection and 20 requests
    #
    delay=`expr $delay + 20 \* $x`
    i=0
    while [ $i -lt $delay ]
    do
	$VERBOSE && $PCP_ECHO_PROG $PCP_ECHO_N ".""$PCP_ECHO_C"
	if $SHOWME
	then
	    echo "+ echo 'connect $1' | $pmlc_prefix pmlc ..."
	    $VERBOSE && echo " done"
	    return 0
	elif echo "connect $1" | eval $pmlc_prefix pmlc 2>&1 | grep "Unable to connect" >/dev/null
	then
            :
        else
            sleep 5
            $VERBOSE && echo " done"
            return 0
        fi

	if _get_pids_by_name pmlogger | grep "^$1\$" >/dev/null
	then
            :
        else
	    $VERBOSE || _message restart
	    echo " process exited!"
	    _check_logfile
	    return 1
	fi
	sleep 5
	i=`expr $i + 5`
    done
    $VERBOSE || _message restart
    echo " timed out waiting!"
    sed -e 's/^/	/' $tmp/out
    _check_logfile
    return 1
}

_message()
{
    case $1
    in
	looking)
	    $PCP_ECHO_PROG $PCP_ECHO_N "Looking for $myname ...""$PCP_ECHO_C"
	    ;;
	get_host)
	    $PCP_ECHO_PROG $PCP_ECHO_N "Getting logged host name from $myname ...""$PCP_ECHO_C"
	    ;;
	get_state)
	    $PCP_ECHO_PROG $PCP_ECHO_N "Contacting $myname to get logging state ...""$PCP_ECHO_C"
	    ;;
	restart)
	    $PCP_ECHO_PROG $PCP_ECHO_N "Waiting for new pmlogger to start ..""$PCP_ECHO_C"
	    ;;
    esac
}

_do_cmd()
{
    if $SHOWME
    then
	echo "+ $1"
    else
	eval $1
    fi
}

# option parsing
#
# pmlogger, without -V version which is redefined as -V (verbose)
# and -s exit_size which is redefined as -s (pmsocks), and ignore
# [ -h host ] and [ -x fd ] as they make no sense in the argument
# part of the pmlogger control file for a long-running pmlogger.
#

ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in

# pmnewlog options and flags
#

	-a)	access="$2"
		shift
		if [ ! -f "$access" ]
		then
		    echo "$prog: Error: cannot find accessfile ($access)"
		    _abandon
		fi
		;;

	-C)	saveconfig="$2"
		shift
		;;

	-N)	SHOWME=true
		CP="echo + cp"
		MV="echo + mv"
		RM="echo + rm"
		KILL="echo + kill"
		;;

	-p)	pid=$2
		shift
		primary=false
		myname="pmlogger (process $pid)"
		connect=$pid
		;;

	-s)	if which pmsocks >/dev/null 2>&1
                then
                    sock_me="pmsocks "
                else
                    echo "$prog: Warning: no pmsocks available, would run without"
                    sock_me=""
                fi
		;;

	-V)	VERBOSE=true
		;;

# pmlogger options and flags that need special handling
#

	-c)	config="$2"
		shift
		if [ ! -f "$config" ]
		then
		    if [ -f "$PCP_VAR_DIR/config/pmlogger/$config" ]
		    then
			config="$PCP_VAR_DIR/config/pmlogger/$config"
		    else
			echo "$prog: Error: cannot find configfile ($config)"
			_abandon
		    fi
		fi
		;;

	-l)	logfile="$2"
		shift
		;;

	-n)	namespace="-n $2"
		args="${args}$1 $2 "
		shift
		;;

	-P)	primary=true
		myname="primary pmlogger"
		;;

# pmlogger flags passed through
#

	-L|-o|-r|-y)
		args="${args}$1 "
		;;

	-D|-H|-K|-m|-t|-T|-v)
		args="${args}$1 $2 "
		shift
		;;

	--)	# end of options, start of arguments
		shift
		break
		;;

	-\?)	pmgetopt --usage --progname=$prog --config=$tmp/usage
		_abandon
		;;
    esac
    shift
done

if [ $# -ne 1 ]
then
    echo "$prog: Insufficient arguments" >&2
    echo >&2
    pmgetopt --usage --progname=$prog --config=$tmp/usage
    _abandon
fi

# initial sanity checking for new archive name
#
archive="$1"

# check that designated pmlogger is really running
#
$VERBOSE && _message looking
$PCP_PS_PROG $PCP_PS_ALL_FLAGS \
| if $primary
then
    grep 'pmlogger .*-P' | grep -v grep
else
    $PCP_AWK_PROG '$2 == '"$pid"' && /pmlogger/ { print }'
fi >$tmp/out

if [ -s $tmp/out ]
then
    # we expect to find only one matching process ...
    #
    if [ "`wc -l <$tmp/out | sed -e 's/ //g'`" -gt 1 ]
    then
	$VERBOSE && echo " botch"
	if $primary
	then
	    echo "$prog: Error: expecting at most one primary pmlogger, found:"
	else
	    echo "$prog: Error: expecting at most one pmlogger with pid=$pid, found:"
	fi
	cat $tmp/out
	_abandon
    fi
    $VERBOSE && echo " found"
    $VERBOSE && cat $tmp/out
    # expecting something like
    # pcp      30019   ... pmlogger ...args... -h hostname ...args...
    # pick pid and hostname (safer to do it here if possible because it
    # captures the possible -h hostname@proxy construct which is lost
    # if one gets the hostname from pmlogger via pmlc)
    #
    pid=`$PCP_AWK_PROG '{ print $2 }' <$tmp/out`
    hostname=`sed -n <$tmp/out -e '/ -h /{
s/.* -h //
s/ .*//
p
}'`
    # may have @proxyhost or @proxyhost:proxyport appended to hostname
    #
    proxyhost=`echo "$hostname" | sed -n -e '/@/{
s/.*@//
s/:.*//
p
}'`
    proxyport=`echo "$hostname" | sed -n -e '/@.*:/s/.*;//p'`
else
    if $VERBOSE
    then
	:
    else
	_message looking
	echo
    fi
    echo "$prog: Error: process not found"
    _abandon
fi

if [ -n "$proxyhost" ]
then
    if [ -n "$proxyport" ]
    then
	pmlc_prefix="PMPROXY_HOST=$proxyhost PMPROXY_PORT=$proxyport"
    else
	pmlc_prefix="PMPROXY_HOST=$proxyhost"
    fi
else
    pmlc_prefix=''
fi

# pass primary/not primary down
#
$primary && args="$args-P "

# pass logfile option down
#
args="$args-l $logfile "

# if not a primary pmlogger, get name of pmcd host pmlogger is connected to
#
if $primary
then
    hostname=localhost
elif [ -z "$hostname" ]
then
    # did not get pmcd hostname from ps output above, talk to pmlogger
    # via pmlc
    #
    # start critical section ... no interrupts due to pmlogger SIGPIPE
    # bug in PCP 1.1
    #
    trap "echo; echo $prog:' Interrupt! ... I am talking to pmlogger, please wait ...'" 1 2 3 15
    $VERBOSE && _message get_host

    _do_cmd "( echo 'connect $connect' ; echo status ) | eval $pmlc_prefix pmlc 2>$tmp/err >$tmp/out"

    # end critical section
    #
    trap "rm -rf $tmp; exit \$status" 0 1 2 3 15

    if $SHOWME || [ ! -s $tmp/err ]
    then
	$VERBOSE && echo " done"
    else
	if grep "Unable to connect" $tmp/err >/dev/null
	then
	    $VERBOSE || _message get_host
	    echo " failed to connect"
	    echo
	    sed -e 's/^/	/' $tmp/err
	    _abandon
	else
	    $VERBOSE || _message get_host
	    echo
	    echo "$prog: Warning: errors from talking to $myname via pmlc"
	    sed -e 's/^/	/' $tmp/err
	    echo
	    echo "continuing ..."
	fi
    fi

    if $SHOWME
    then
	hostname=somehost
    else
	hostname=`sed -n -e '/^pmlogger/s/.* from host //p' <$tmp/out`
	if [ -z "$hostname" ]
	then
	    echo "$prog: Error: failed to get host name from $myname"
	    echo "This is what was collected from $myname."
	    echo
	    sed -e 's/^/	/' $tmp/out
	    _abandon
	fi
	args="$args-h $hostname "
    fi
else
    # got hostname from ps output
    #
    args="$args-h $hostname "
fi

# extract/construct config file if required
#
if [ -z "$config" ]
then
    # start critical section ... no interrupts due to pmlogger SIGPIPE
    # bug in PCP 1.1
    #
    trap "echo; echo $prog:' Interrupt! ... I am talking to pmlogger, please wait ...'" 1 2 3 15
    $VERBOSE && _message get_state

    # iterate over top-level names in pmns, and query pmlc for
    # current configuration ... note exclusion of "proc" metrics
    # ... others may be excluded in a similar fashion
    #
    if $SHOWME
    then
	echo "+ ( echo 'connect $connect'; echo 'query ...'; ... ) | eval $pmlc_prefix pmlc $namespace | $PCP_AWK_PROG ..."
    else
	echo "connect $connect" >$tmp/pmlc.cmd
	for top in `pminfo -h $hostname $namespace \
                    | sed -e 's/\..*//' -e '/^proc$/d' -e '/^hotproc$/d'\
		    | sort -u`
	do
	    echo "query $top" >>$tmp/pmlc.cmd
	done
	eval $pmlc_prefix pmlc $namespace <$tmp/pmlc.cmd 2>$tmp/err \
	| $PCP_AWK_PROG >$tmp/out '
/^[^ ]/						{ metric = $1; next }
$1 == "mand" || ( $1 == "adv" && $2 == "on" ) 	{ print $0 " " metric }'
    fi

    # end critical section
    #
    trap "rm -rf $tmp; exit \$status" 0 1 2 3 15

    if $SHOWME || [ ! -s $tmp/err ]
    then
	$VERBOSE && echo " done"
    else
	if grep "Unable to connect" $tmp/err >/dev/null
	then
	    $VERBOSE || _message get_state
	    echo " failed to connect"
	    echo
	    sed -e 's/^/	/' $tmp/err
	    _abandon
	else
	    $VERBOSE || _message get_state
	    echo
	    echo "$prog: Warning: errors from talking to $myname via pmlc"
	    sed -e 's/^/	/' $tmp/err
	    echo
	    echo "continuing ..."
	fi
    fi

    if [ ! -s $tmp/out ]
    then
	if $SHOWME
	then
	    :
	else
	    echo "$prog: Error: failed to collect configuration info from $myname"
	    echo "Most likely this pmlogger instance is inactive."
	    _abandon
	fi
    fi

    # convert to a pmlogger config file
    #
    if $SHOWME
    then
	echo "+ create new pmlogger config file ..."
    else
	sed <$tmp/out >$tmp/config \
	    -e 's/ on  nl/ on/' \
	    -e 's/ off nl/ off/'\
	    -e 's/  *mand  *\(o[nf]*\) /log mandatory \1 /' \
	    -e 's/  *adv  *on /log advisory on/' \
	    -e 's/\[[0-9][0-9]* or /[/' \
	    -e 's/\(\[[^]]*]\) \([^ ]*\)/\2 \1/' \
	    -e 's/   */ /g'

	if [ ! -s $tmp/config ]
	then
	    echo "$prog: Error: failed to generate a pmlogger configuration file for pmlogger"
	    echo "This is what was collected from $myname."
	    echo
	    sed -e 's/^/	/' $tmp/out
	    _abandon
	fi
    fi
    config=$tmp/config
fi

# optionally append access control specifications
#
if [ -n "$access" ]
then
    if grep '\[access]' $config >/dev/null
    then
	echo "$prog: Error: pmlogger configuration file already contains an"
	echo "	access control section, specifications from \"$access\" cannot"
	echo "	be applied."
	_abandon
    fi
    cat $access >>$config
fi

# add config file to the args, save config file if -C
#
args="$args-c $config "
if [ -n "$saveconfig" ]
then
    if eval $CP $config $saveconfig
    then
	echo "New pmlogger configuration file saved as $saveconfig"
    else
	echo "$prog: Warning: unable to save configuration file as $saveconfig"
    fi
fi

# kill off existing pmlogger
#
$VERBOSE && $PCP_ECHO_PROG $PCP_ECHO_N "Terminating $myname ...""$PCP_ECHO_C"
for sig in USR1 TERM KILL
do
    $VERBOSE && $PCP_ECHO_PROG $PCP_ECHO_N " SIG$sig ...""$PCP_ECHO_C"
    eval $KILL -s $sig $pid
    sleep 5
    [ "`_check_pid $pid`" = "" ] && break
done

if [ "`_check_pid $pid`" = "" ]
then
    $VERBOSE && echo " done"
else
    echo " failed!"
    _abandon
fi

# the archive folio Latest is for the most recent archive in this directory
#
dir=`dirname $archive`
eval $RM -f $dir/Latest

# clean up port-map, just in case, and prepare primary pmlogger environment
#
PM_LOG_PORT_DIR="$PCP_TMP_DIR/pmlogger"
eval $RM -f "$PM_LOG_PORT_DIR"/$pid
if $primary
then
    eval $RM -f "$PM_LOG_PORT_DIR/primary"
    envs=`grep ^PMLOGGER "$PMLOGGERENVS" 2>/dev/null`
else
    envs=""
fi

# finally do it, ...
#
cd $dir
$SHOWME && echo "+ cd $dir"
[ "$dir" = . ] && dir=`pwd`
archive=`basename $archive`

# handle duplicates/aliases (happens when pmlogger is restarted
# within a minute and basename is the same)
#
suff=''
for file in $archive.*
do
    [ "$file" = "$archive"'.*' ] && continue
    # we have a clash! ... find a new -number suffix for the
    # existing files ... we are going to keep $archive for the
    # new pmlogger below
    #
    if [ -z "$suff" ]
    then
	for xx in 0 1 2 3 4 5 6 7 8 9
	do
	    for yy in 0 1 2 3 4 5 6 7 8 9
	    do
		[ "`echo $archive-${xx}${yy}.*`" != "$archive-${xx}${yy}.*" ] && continue
		suff=${xx}$yy
		break
	    done
	    [ ! -z "$suff" ] && break
	done
	if [ -z "$suff" ]
	then
	    echo "$prog: Error: unable to break duplicate clash for archive basename \"$archive\""
	    _abandon
	fi
	$VERBOSE && echo "Duplicate archive basename ... rename $archive.* files to $archive-$suff.*"
	[ -f SaveLogs/$archive.log ] && eval $MV SaveLogs/$archive.log SaveLogs/$archive-$suff.log
    fi
    eval $MV -f $file `echo $file | sed -e "s/$archive/&-$suff/"`
done

$VERBOSE && echo "Launching new pmlogger in directory \"$dir\" as ..."
[ -f $logfile ] && eval $MV -f $logfile $logfile.prior
$VERBOSE && echo "${sock_me}$PMLOGGER $args$archive"

if $SHOWME
then
    echo "+ ${sock_me}$PMLOGGER $args$archive &"
    echo "+ ... assume pid is 12345"
    new_pid=12345
else
    eval $envs '${sock_me}$PMLOGGER $args$archive &'
    new_pid=$!
fi

# stall a bit ...
#
STALL_TIME=10
sleep $STALL_TIME

$VERBOSE && _message restart
if _check_logger $new_pid || $SHOWME
then
    $VERBOSE && echo "New pmlogger status ..."
    $VERBOSE && _do_cmd "( echo 'connect $new_pid'; echo status ) | $pmlc_prefix pmlc"

    # make the "Latest" archive folio
    #
    i=0
    failed=true
    WAIT_TIME=10
    while [ $i -lt $WAIT_TIME ]
    do
	if $SHOWME || [ -f $archive.0 -a -f $archive.meta -a $archive.index ]
	then
	    _do_cmd "mkaf $archive.0 >Latest" 2>$tmp/err
	    if [ -s $tmp/err ]
	    then
		# errors from mkaf typically result from race conditions
		# at the start of pmlogger, e.g.
		# Warning: cannot extract hostname from archive "..." ...
		#
		# simply keep trying
		:
	    else
		failed=false
		break
	    fi
	fi
	sleep 1
	i=`expr $i + 1`
    done

    if $failed
    then
	ELAPSED=`expr $STALL_TIME + $WAIT_TIME`
	echo "Warning: pmlogger [pid=$new_pid host=$hostname] failed to create archive files within $ELAPSED seconds"
	if [ -f $tmp/err ]
	then
	    echo "Warnings/errors from mkaf ..."
	    cat $tmp/err
	fi
    fi
else
    _abandon
fi

# if SaveLogs exists in the same directory that the archive
# is being created, save pmlogger log file there as well
# (we've already cd'd into $dir)
#
if [ -d SaveLogs ]
then
    if [ ! -f SaveLogs/$LOGNAME.log ]
    then
	_do_cmd "ln $logfile SaveLogs/$archive.log"
    fi
fi

exit
